mathematica/integrator_plots.cpp

```diff
@@ -219,137 +328,337 @@ class WorkErrorGraphGenerator {
     problem.equation.compute_acceleration = std::bind(
         compute_accelerations_, _1, _2, _3, &number_of_evaluations);
     problem.initial_state = initial_state_;
+    first_order_problem.equation.compute_derivative =
+        [&number_of_evaluations, this](
+            Instant const& s,
+            std::tuple<Position<World>, Velocity<World>> const& y,
+            std::tuple<Velocity<World>, Vector<Acceleration, World>>& yʹ) {
+          std::vector<Vector<Acceleration, World>> acceleration(1, {});
+          auto const& [q, v] = y;
+          auto const status = compute_accelerations_(
+              s, {q}, acceleration, &number_of_evaluations);
+          yʹ = {v, acceleration[0]};
+          return status;
+        };
+    first_order_problem.initial_state.s = initial_state_.time;
+    first_order_problem.initial_state.y = {initial_state_.positions[0],
+                                           initial_state_.velocities[0]};
     auto const t0 = problem.initial_state.time.value;
     Length max_q_error;
     Speed max_v_error;
     Energy max_e_error;
     auto append_state = [this, &max_q_error, &max_v_error, &max_e_error](
-      ODE::State const& state) {
+        SecondOrderODE::State const& state) {
       auto const errors = compute_errors_(state);
       max_q_error = std::max(max_q_error, errors.q_error);
       max_v_error = std::max(max_v_error, errors.v_error);
       max_e_error = std::max(max_e_error, errors.e_error);
     };
-    Time const Δt = method.evaluations * starting_step_size_per_evaluation_ /
-                    std::pow(step_reduction_, time_step_index);
-    auto const instance =
-        method.integrator.NewInstance(problem, append_state, Δt);
-
-    CHECK_OK(instance->Solve(tmax_));
-    // Log both the actual number of evaluations and a theoretical number
-    // that ignores any startup costs; that theoretical number is the one
-    // used for plotting.
-    int const amortized_evaluations =
-        method.evaluations * static_cast<int>(std::floor((tmax_ - t0) / Δt));
-    LOG_EVERY_N(INFO, 50) << "[" << method_index << "," << time_step_index
-                          << "] " << problem_name_ << ": "
-                          << number_of_evaluations << " actual evaluations ("
-                          << amortized_evaluations << " amortized) with "
-                          << method.name;
-    // We plot the maximum error, i.e., the L∞ norm of the error.
-    // [BM02] or [BCR01a] tend to use the average error (the normalized L¹ norm)
-    // instead.
-    q_errors_[method_index][time_step_index] = max_q_error;
-    v_errors_[method_index][time_step_index] = max_v_error;
-    e_errors_[method_index][time_step_index] = max_e_error;
-    evaluations_[method_index][time_step_index] = amortized_evaluations;
+    auto append_first_order_state =
+    [this, &max_q_error, &max_v_error, &max_e_error](
+        FirstOrderODE::State const& state) {
+      SecondOrderODE::State second_order_state;
```

Comment:
Use aggregate initialization to fill this struct.

Reply:
State is not an aggregate type, because it has a user-declared constructor (that lengthens to double precision). Here we have double precision already (from the other state).

```C++
  struct State final {
    // […]
    State(IndependentVariable const& t,
          DependentVariables const& q,
          DependentVariableDerivatives const& v);

    DoublePrecision<IndependentVariable> time;
    std::vector<DoublePrecision<DependentVariable>> positions;
    std::vector<
        DoublePrecision<Derivative<DependentVariable, IndependentVariable>>>
        velocities;
    // […]
  };
```